google/cloud/internal/openssl_util.cc

```diff
@@ -16,118 +16,142 @@
           // a password, which we don't currently support.
-          nullptr),
-      &EVP_PKEY_free);
+          nullptr));
   if (!private_key) {
     return Status(StatusCode::kInvalidArgument,
-                  "Invalid ServiceAccountCredentials: "
-                  "could not parse PEM to get private key ");
+                  "Invalid ServiceAccountCredentials - "
+                  "could not parse PEM to get private key: " +
+                      CaptureSslErrors());
+  }
+
+  auto digest_ctx = GetDigestCtx();
+  if (!digest_ctx) {
+    return Status(StatusCode::kInvalidArgument,
+                  "Invalid ServiceAccountCredentials - "
+                  "could not create context for OpenSSL digest: " +
+                      CaptureSslErrors());
   }
 
-  int const digest_sign_success_code = 1;
-  if (digest_sign_success_code !=
-      EVP_DigestSignInit(digest_ctx.get(),
-                         nullptr,  // `EVP_PKEY_CTX **pctx`
-                         digest_type,
-                         nullptr,  // `ENGINE *e`
-                         private_key.get())) {
+  auto constexpr kOpenSslSuccess = 1;
+  if (EVP_DigestSignInit(digest_ctx.get(), nullptr, EVP_sha256(), nullptr,
+                         private_key.get()) != kOpenSslSuccess) {
     return Status(StatusCode::kInvalidArgument,
-                  "Invalid ServiceAccountCredentials: "
-                  "could not initialize PEM digest. ");
+                  "Invalid ServiceAccountCredentials - "
+                  "could not initialize signing digest: " +
+                      CaptureSslErrors());
   }
 
-  if (digest_sign_success_code !=
-      EVP_DigestSignUpdate(digest_ctx.get(), str.data(), str.length())) {
+  if (EVP_DigestSignUpdate(digest_ctx.get(), str.data(), str.size()) !=
+      kOpenSslSuccess) {
     return Status(StatusCode::kInvalidArgument,
-                  "Invalid ServiceAccountCredentials: "
-                  "could not update PEM digest. ");
+                  "Invalid ServiceAccountCredentials - could not sign blob: " +
+                      CaptureSslErrors());
   }
 
-  std::size_t signed_str_size = 0;
-  // Calling this method with a nullptr buffer will populate our size var
-  // with the resulting buffer's size. This allows us to then call it again,
-  // with the correct buffer and size, which actually populates the buffer.
-  if (digest_sign_success_code !=
-      EVP_DigestSignFinal(digest_ctx.get(),
-                          nullptr,  // unsigned char *sig
-                          &signed_str_size)) {
+  // The OpenSSL "C" APIs uses `size_t`, I (coryan) believe all platforms we
+  // support satisfy this assertion, and suspect this is true in all platforms.
+  // However, if the code ever finds its way to one where it is not true, it
+  // would be better to get a compile-time error.
+  static_assert(std::is_same<std::size_t, size_t>::value,
+                "Expect std::size_t == size_t");
```

Comment:
The C++ standard requires that ::size_t and std::size_t are identical.

Reply:
I suspected that would be the case, but TIL.